### PR TITLE
MERL-484: telemetry

### PR DESCRIPTION
**Pull request by matthewguywright**
## Description

This PR introduces a `faust` CLI that acts as a pass through to the `next` CLI, however, with our own telemetry that runs prior to the `next` CLI.

When running the `faust` CLI for the first time, you are asked if you would like to enroll in Faust Telemetry. By responding with either `true` or `false`, your preferences are then saved to your local machine (`~/.config/configstore/faust.json` on Mac) as a JSON file:

```jsonc
{
  telemetry: {
    notifiedAt: '12345', // timestamp the user was first notified about Faust telemetry
    anonymousId: '0000-0000-0000', // An anonymous v4 UUID for GA
    enabled: true // Either true or false based on user preferences
  }
}
```

From there, any time the `faust` CLI is ran, we check the users' preferences, and if telemetry is enabled, we send a GA event with relevant information.

In addition to the basic Next CLI commands like `faust start`, `faust dev`, and `faust build`, there is also a new command `faust faust-telemetry`, which allows the user to modify their telemetry preferences.

## Testing

1. Checkout PR branch
2. Uncomment this line: https://github.com/wpengine/faustjs/blob/61bee88bb152e00c206de939aa092f7558bd5313/packages/faustwp-cli/index.ts#L90
3. Run `npm run build -w @faustwp/core` from the monorepo root
4. Run `npm run build -w @faustwp/cli` from the monorepo root
5. Run `npm run dev -w @faustwp/getting-started-example` to start the Faust example app in dev mode
6. Notice the Telemetry prompt
7. Type `true`
8. You should now see a telemetry event being `console.log`'d
11. Now, run `npx faustnx faustnx-telemetry` and change your preferences to `false`
12. Run `npm run build -w faust-nx-getting-started` notice that the telemetry event is not being `console.log`'d

## Telemetry Data

The following data is what is beign sent to GA.

```ts
export interface TelemetryData {
  setting_has_frontend_uri?: boolean;
  setting_redirects_enabled?: boolean;
  setting_rewrites_enabled?: boolean;
  setting_themes_disabled?: boolean;
  setting_img_src_replacement_enabled?: boolean;
  faustwp_version?: string;
  is_wpe?: boolean;
  multisite?: boolean;
  php_version?: string;
  wp_version?: string;
  node_faustwp_core_version?: string;
  node_faustwp_cli_version?: string;
  node_apollo_client_version?: string;
  node_version?: string;
  node_next_version?: string;
  node_is_development?: boolean;
}
```

### CODE REVIEW

File: packages/faustwp-cli/index.ts [hunk 1]
```diff
@@ -1,15 +1,77 @@
...
 
+import Configstore from 'configstore';
 import { spawn } from 'child_process';
 import dotenv from 'dotenv-flow';
-
-import { getCliArgs, validateFaustEnvVars } from './utils/index.js';
+import {
+  marshallTelemetryData,
+  getCliArgs,
+  validateFaustNXEnvVars,
```
Comment (theodesp) [inline packages/faustwp-cli/index.ts:9]: This should be `validateFaustEnvVars`

File: packages/faustwp-cli/index.ts [hunk 2]
```diff
@@ -1,15 +1,77 @@
...
+  promptUserForTelemetryPref,
+  sendTelemetryData,
+  requestWPTelemetryData,
+  noticeLog,
+} from './utils/index.js';
 
 dotenv.config();
-validateFaustEnvVars();
+const CONFIG_STORE_NAME = 'faustnx';
```
Comment (theodesp) [inline packages/faustwp-cli/index.ts:17]:
```suggestion
const CONFIG_STORE_NAME = 'faust';
```

File: packages/faustwp-cli/index.ts [hunk 3]
```diff
@@ -1,15 +1,77 @@
...
+   */
+  if (
+    !config.all?.telemetry ||
+    config.all?.telemetry?.enabled === undefined ||
+    !config.all?.telemetry?.anonymousId
+  ) {
+    await promptUserForTelemetryPref(true, config);
+  }
+
+  if (getCliArgs()[0] === 'faustnx-telemetry') {
```
Comment (theodesp) [inline packages/faustwp-cli/index.ts:34]:
```suggestion
  if (getCliArgs()[0] === 'faust-telemetry') {
```

File: packages/faustwp-cli/index.ts [hunk 4]
```diff
@@ -1,15 +1,77 @@
...
+    process.exit(0);
+  }
+
+  validateFaustNXEnvVars();
+
+  const shouldFireTelemetryEvent =
+    (getCliArgs()[0] === 'dev' || getCliArgs()[0] === 'build') &&
+    config.get('telemetry.enabled') === true &&
+    config.get('telemetry.anonymousId') &&
+    process.env.HEADLESS_SECRET_KEY;
```
Comment (theodesp) [inline packages/faustwp-cli/index.ts:46]:
```suggestion
    process.env.FAUSTWP_SECRET_KEY;
```

File: packages/faustwp-cli/index.ts [hunk 5]
```diff
@@ -1,15 +1,77 @@
...
+
+  if (shouldFireTelemetryEvent) {
+    try {
+      const wpTelemetryData = await requestWPTelemetryData(
+        process.env.NEXT_PUBLIC_WORDPRESS_URL!,
+        process.env.HEADLESS_SECRET_KEY!,
```
Comment (theodesp) [inline packages/faustwp-cli/index.ts:52]:
```suggestion
        process.env.FAUSTWP_SECRET_KEY!,
```

File: packages/faustwp-cli/utils/marshallTelemetryData.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      image_source_replacement_enabled?: boolean;
+    };
+    faustwp_version?: string;
+    is_wpe?: boolean;
+    multisite?: boolean;
+    php_version?: string;
+    wp_version?: string;
+  };
+  node: {
+    faustnx_version?: string;
```
Comment (theodesp) [inline packages/faustwp-cli/utils/marshallTelemetryData.ts:21]:
```suggestion
    faust_version?: string;
```

File: packages/faustwp-cli/utils/marshallTelemetryData.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    faustnx_cli_version?: string;
```
Comment (theodesp) [inline packages/faustwp-cli/utils/marshallTelemetryData.ts:22]:
```suggestion
    faust_cli_version?: string;
```

File: packages/faustwp-cli/utils/marshallTelemetryData.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+/**
+ * Marshall the JS and WP telemetry data into a single object for GA.
+ * @param wpTelemetryData
+ */
+export const marshallTelemetryData = (
+  wpTelemetryData: WPTelemetryResponseData | undefined,
+): TelemetryData => {
+  const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
+
+  let telemetryData: TelemetryData = {
```
Comment (theodesp) [inline packages/faustwp-cli/utils/marshallTelemetryData.ts:65]: Change to correct library names here

File: packages/faustwp-cli/utils/promptUserForTelemetryPref.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ */
+export const promptUserForTelemetryPref = async (
+  isInit: boolean,
+  config: Configstore,
+) => {
+  const { isTelemetryEnabled } = await prompt.get({
+    properties: {
+      isTelemetryEnabled: {
+        description:
+          'Would you like to enable telemetry in FaustNX? This helps us make more informed feature decisions. true/false',
```
Comment (theodesp) [inline packages/faustwp-cli/utils/promptUserForTelemetryPref.ts:21]:
```suggestion
          'Would you like to enable telemetry in Faust? This helps us make more informed feature decisions. true/false',
```

File: packages/faustwp-cli/utils/sendTelemetryData.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import fetch from 'node-fetch';
+import { TelemetryData } from './marshallTelemetryData';
+
+const GA_TRACKING_ENDPOINT = 'http://www.google-analytics.com/debug/collect';
+const GA_TRACKING_ID = 'GA-xxxx';
```
Comment (theodesp) [inline packages/faustwp-cli/utils/sendTelemetryData.ts:5]: Secret key is missing

File: internal/faustjs.org/docs/telemetry.mdx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+description: We use Telemetry in Faust.js to help us understand how users use our projects.
+---
+
+Faust.js has opt-in telemetry for both the FaustWP plugin and FaustNX package. The plugin and package both collect information to help us make more informed product decisions.
+
+## What is being collected?
+
+Once you opt-in, faust will collect the following information at build time:
+
+### FaustNX
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:13]:
```suggestion
### Faust
```

File: internal/faustjs.org/docs/telemetry.mdx [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+- FaustNX version
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:15]:
```suggestion
- Faust version
```

File: internal/faustjs.org/docs/telemetry.mdx [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+:::info
+We **do not** collect any personally identifiable information or sensitive information like environment variables, file paths, or any application URLs.
+:::
+
+## How do I enable/disable telemetry?
+
+Telemetry in Faust.js is opt-in by default, meaning we won't collect any telemetry information unless you explicitly enable it.
+
+If this is your first time using the `faustnx` cli, you will be prompted for your telemetry preferences:
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:38]:
```suggestion
If this is your first time using the `faustwp` cli, you will be prompted for your telemetry preferences:
```

File: packages/faustwp-cli/utils/marshallTelemetryData.ts [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+      image_source_replacement_enabled?: boolean;
+    };
+    faustwp_version?: string;
+    is_wpe?: boolean;
+    multisite?: boolean;
+    php_version?: string;
+    wp_version?: string;
+  };
+  node: {
+    faust_version?: string;
```
Comment (blakewilson) [inline packages/faustwp-cli/utils/marshallTelemetryData.ts:21]: We may want to rename this to `faust_core_version`

```suggestion
    faust_core_version?: string;
```

File: packages/faustwp-cli/utils/marshallTelemetryData.ts [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+ */
+export const marshallTelemetryData = (
+  wpTelemetryData: WPTelemetryResponseData | undefined,
+): TelemetryData => {
+  const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
+
+  let telemetryData: TelemetryData = {
+    node: {
+      faustnx_version: sanitizePackageJsonVersion(
+        packageJson?.dependencies?.['faust-nx'],
```
Comment (blakewilson) [inline packages/faustwp-cli/utils/marshallTelemetryData.ts:68]:
```suggestion
      faust_core_version: sanitizePackageJsonVersion(
        packageJson?.dependencies?.['@faustwp/core'],
```

File: packages/faustwp-cli/utils/marshallTelemetryData.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+      ),
+      faustnx_cli_version: sanitizePackageJsonVersion(
+        packageJson?.dependencies?.['faust-nx-cli'],
```
Comment (blakewilson) [inline packages/faustwp-cli/utils/marshallTelemetryData.ts:71]:
```suggestion
      faust_cli_version: sanitizePackageJsonVersion(
        packageJson?.dependencies?.['@faustwp/cli'],
```

File: internal/faustjs.org/docs/telemetry.mdx [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  <img
+    src="/docs/img/telemetry-cli-prompt.png"
+    alt="Screenshot of the Telemetry prompt in the FaustNX package"
+  />
+</a>
+
+Once your preferences have been set, you can run the following command to see the prompt again to change your preferences:
+
+```sh
+npx faustnx faustnx-telemetry
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:50]:
```suggestion
npx faustwp faust-telemetry
```

**Comment by theodesp** [comment]
The demo Image needs replacement as it shows usage of `faustnx`:

internal/faustjs.org/static/docs/img/telemetry-cli-prompt.png

File: internal/faustjs.org/docs/telemetry.mdx [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+## How do I enable/disable telemetry?
+
+Telemetry in Faust.js is opt-in by default, meaning we won't collect any telemetry information unless you explicitly enable it.
+
+If this is your first time using the `faustwp` cli, you will be prompted for your telemetry preferences:
+
+<a href="/docs/img/telemetry-cli-prompt.png" target="_blank">
+  <img
+    src="/docs/img/telemetry-cli-prompt.png"
+    alt="Screenshot of the Telemetry prompt in the FaustNX package"
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:43]:
```suggestion
    alt="Screenshot of the Telemetry prompt in the Faust package"
```

File: internal/faustjs.org/docs/telemetry.mdx [hunk 19]
```diff
@@ -0,0 +1,51 @@
+---
+slug: telemetry
+title: Telemetry
+description: We use Telemetry in Faust.js to help us understand how users use our projects.
+---
+
+Faust.js has opt-in telemetry for both the FaustWP plugin and FaustNX package. The plugin and package both collect information to help us make more informed product decisions.
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:7]:
```suggestion
Faust.js has opt-in telemetry for both the FaustWP plugin and Faust package. The plugin and package both collect information to help us make more informed product decisions.
```

File: packages/faustwp-cli/utils/sendTelemetryData.ts [hunk 20]
```diff
@@ -0,0 +1,47 @@
+import fetch from 'node-fetch';
+import { TelemetryData } from './marshallTelemetryData';
+
+const GA_TRACKING_ENDPOINT = 'http://www.google-analytics.com/debug/collect';
+const GA_TRACKING_ID = 'G-KPVSTHK1G4';
```
Comment (theodesp) [inline packages/faustwp-cli/utils/sendTelemetryData.ts:5]: You also need to include the secret query parameter as well:

https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference?client_type=firebase

Here is an example:
https://github.com/nkowaokwu/igbo_api/blob/81581a20c2a1556b28a66d6114370f2809620fb5/src/middleware/analytics.js#L33-L36

File: internal/faustjs.org/docs/telemetry.mdx [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+
+:::info
+We **do not** collect any personally identifiable information or sensitive information like environment variables, file paths, or any application URLs.
+:::
+
+## How do I enable/disable telemetry?
+
+Telemetry in Faust.js is opt-in by default, meaning we won't collect any telemetry information unless you explicitly enable it.
+
+If this is your first time using the `faustwp` cli, you will be prompted for your telemetry preferences:
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:38]:
```suggestion
If this is your first time using the `faust` cli, you will be prompted for your telemetry preferences:
```

Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:38]:
```suggestion
If this is your first time using the `faust` cli, you will be prompted for your telemetry preferences:
```

**Comment by theodesp** [comment]
Conflicts again...

File: internal/faustjs.org/docs/telemetry.mdx [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+### Faust
+
+- Faust version
+- Next.js version
+- Node.js version
+- Apollo Client version
+- If the command is running in a development environment
+
+### FaustWP
```
Comment (theodesp) [inline internal/faustjs.org/docs/telemetry.mdx:21]:
```suggestion
### Faust Plugin
```

**Review by theodesp — APPROVED** [review]

**Comment by theodesp** [comment]
Awesome!

**Comment by blakewilson** [comment]
@matthewguywright Please wait to merge this until we do our release today.

**Review by theodesp — APPROVED** [review]

**Pull request merged**